File: ts/src/util/crc32.ts
// CRC32 checksum implementation (IEEE polynomial)

const TABLE = new Uint32Array(256);

for (let i = 0; i < 256; i++) {
  let crc = i;
  for (let j = 0; j < 8; j++) {
    crc = crc & 1 ? (crc >>> 1) ^ 0xedb88320 : crc >>> 1;
  }
  TABLE[i] = crc >>> 0;
}

export function crc32(data: Uint8Array): number {
  let crc = 0xffffffff;
  for (let i = 0; i < data.byteLength; i++) {
    crc = TABLE[(crc ^ data[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

export function crc32Update(prev: number, data: Uint8Array): number {
  let crc = prev ^ 0xffffffff;
  for (let i = 0; i < data.byteLength; i++) {
    crc = TABLE[(crc ^ data[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

File: internal/util/crc32.go
package util

import "hash/crc32"

// IEEE polynomial CRC32, same as zlib.
var crc32Table = crc32.MakeTable(crc32.IEEE)

func CRC32(data []byte) uint32 {
	return crc32.Checksum(data, crc32Table)
}

func CRC32Update(prev uint32, data []byte) uint32 {
	return crc32.Update(prev, crc32Table, data)
}
